### PR TITLE
TUF Python Client Example/Tutorial

### PR DESCRIPTION
**Pull request by kairoaraujo**
It is a simple example of TUF ngclient implementation.

This example contains a README.rst that is a tutorial/how-to-use
this simple client using static test data from TUF repository.

The code aims to be straightforward implementation, using basic
concepts from Python and Command Line Interface.

This is part of theupdateframework#1518

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Description of the changes being introduced by the pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [X] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [X] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by kairoaraujo** [comment]
I am still tagging it as WIP to improve in general.
The code uses basic concepts about Python. I avoided OO models and implemented the minimum number of functions to be more sequential and easier to understand. Is it a good approach?

The ``client_example`` is a subfolder in ``examples``, and the ``README.rst`` can be visible as the tutorial for it, any better suggestion?

**Review by jku — COMMENTED** [review]
nice, thanks. Some comments (some of these could be left for future work):
* I think argparse is probably a good choice: it's verbose but part of the base library and will allow multiple subcommands without making a mess
* init should be implicit, I think (as in it should always happen)? it's the correct example for real clients and as long as the printed output is valid it should never be a problem either (also if init is implicit, then it can probably be combined with tuf_updater)
* I would not use "[INFO]" and "[ERROR]" like this: to me it implies you are using logging and that I can decrease/increase verbosity but I can't, it's just print()... maybe you should be using logging (and set default level to INFO or something)?
* linting: we want this code linted (with the same config that ngclient uses)
* docs: I think in the end the documentation should get included in the published docs (probably under a new top level "examples" title)... this does not have to happen right now but I think we should delay reviewing and polishing the doc until we see the published form. My gut feeling is that we should _not_ include details like how to clone the repo or where to find the example code etc: these are things that change over time and then someone would need to update them. We have a "instructions for contributors" document, let's link to that if we really have to but let's not repeat ourselves.

File: examples/client_example/client_example.py
Comment (lukpueh) [inline examples/client_example/client_example.py]: ?

**Review by lukpueh — COMMENTED** [review]
Cool stuff! I suggest we start running pylint/black/isort over code in the new examples directory too. Can you add the directory to the `lint` directive in `tox.ini`?

Also, I think the printed info text could be slightly improved (I say this as a non-native English speaker). :)

File: examples/client_example/README.rst
Comment (lukpueh) [inline examples/client_example/README.rst]: Any reason to not use Markdown?

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: @lukpueh, Friday evening, I did use a piece of my code for ``argparse``.

File: examples/client_example/README.rst
Comment (kairoaraujo) [reply examples/client_example/README.rst]: No reason to use Markdown, actually I did start with Markdown, but I thought it could be used with the RTD/Sphinx.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I have to say that English is the hard part for me here. 😅

File: examples/client_example/README.rst
Comment (lukpueh) [reply examples/client_example/README.rst]: RTD can do Markdown too.

**Comment by lukpueh** [comment]
> I have to say that English is the hard part for me here. 😅

Maybe we can ask our technical writer and chief of words @jhdalek55 for help with that.

**Comment by kairoaraujo** [comment]
> nice, thanks. Some comments (some of these could be left for future work):
> 
> * I think argparse is probably a good choice: it's verbose but part of the base library and will allow multiple subcommands without making a mess 

Yes, there are more fancy libraries, but I think argparse will do the work without extra dependencies.

> * init should be implicit, I think (as in it should always happen)? it's the correct example for real clients and as long as the printed output is valid it should never be a problem either (also if init is implicit, then it can probably be combined with tuf_updater)

Yes, indeed. I was in doubt about which direction to choose.

> * I would not use "[INFO]" and "[ERROR]" like this: to me it implies you are using logging and that I can decrease/increase verbosity but I can't, it's just print()... maybe you should be using logging (and set default level to INFO or something)?

Actually, I started with logging, but I tried to make it less complex in the sense most of the basic tutorials try to use "print".
I'm open to both.

> * linting: we want this code linted (with the same config that ngclient uses) 

👍 

> * docs: I think in the end the documentation should get included in the published docs (probably under a new top level "examples" title)... this does not have to happen right now but I think we should delay reviewing and polishing the doc until we see the published form. My gut feeling is that we should _not_ include details like how to clone the repo or where to find the example code etc: these are things that change over time and then someone would need to update them. We have a "instructions for contributors" document, let's link to that if we really have to but let's not repeat ourselves.

I understand that we do not want to repeat ourselves, making it more maintainable, and I agree.
I think the challenge here is to give to who land to the example tutorial a non-segregated experience. I mean, something that you can follow in a single experience and not switch across different docs.

File: examples/client_example/README.rst
Comment (kairoaraujo) [reply examples/client_example/README.rst]: Should we agree to move to Markdown?

**Comment by jhdalek55** [comment]
@jhdalek55 ...I'm happy to work on any English language issues. Just let me know what needs to be reviewed and when. I generally prefer to hold off on reviews until the technical content is more or less confirmed.

**Comment by jku** [comment]
> Actually, I started with logging, but I tried to make it less complex in the sense most of the basic tutorials try to use "print".
I'm open to both.

Same, either way works for me.

On the question of "is this good for a tutorial/example?", I think the goals should be
* example that works out of the box
* example code that shows how to write good code that uses python-tuf 
* tutorial material

_in this order_: my opinion is that good example code is more important than a tutorial that literally anyone could follow.

**Comment by kairoaraujo** [comment]
The last commit addresses more general aspects such as lint, README format, etc.
I will work on some designs improvement mentioned by @jku , so still WIP. 👨🏽‍💻

**Comment by kairoaraujo** [comment]
Some questions:
- Should we remove the README.md and make the approach used in the https://github.com/theupdateframework/python-tuf/pull/1685?
- Should we add to the example code more TUF exceptions handling

**Comment by lukpueh** [comment]
> * Should we remove the README.md and make the approach used in the [Docs: Add repository tutorial based on metadata API  (WIP) #1685](https://github.com/theupdateframework/python-tuf/pull/1685)?

Let me finalize my example in that PR and gather some more feedback. I'm still unsure, if a markdown wouldn't be more appropriate, in terms of making the accompanying text more readable. IMO the question boils down to readability vs. ease of testing. (see https://github.com/theupdateframework/python-tuf/issues/808#issuecomment-442155081 and https://github.com/in-toto/demo/pull/7 for related discussion about markdown doc testing)

**Comment by jku** [comment]
> * Should we remove the README.md and make the approach used in the [Docs: Add repository tutorial based on metadata API #1685](https://github.com/theupdateframework/python-tuf/pull/1685)?

I think your code and 1685 are different for two reasons: 
1. this code can be, or can become, a "real" example application. The metadata API tutorial is a stopgap solution until we have a repository library
2. the client side is "well defined": all clients will operate in much the same way so it is possible for us to define the "canonical use pattern". Repository code on the other hand might be wildly different (think CLI apps vs Warehouse)

So I don't think there's a need for these two solutions to look similar.

> * Should we add to the example code more TUF exceptions handling

For the client app my thinking has been that it should need to handle RepositoryError only (although I see you had to handle some requests issues as well -- we can consider those after this merges): RepositoryError means the update cannot continue because of an issue with TUF metadata: there could be many specific reasons but the client code cannot possibly deal with any of them.

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: Do we really need an explicit init?

Couldn't the app always initialize if the expected location does not contain root.json? As long as the logging/output is clear about what it is doing it should be very clear to user: "Found trusted root in {dir}" vs "Bootstrapping trusted root in {dir}"

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: argparse is not a great API but I still like using the namespace object more than downgrading it to a dict:

```suggestion
    args = client_args.parse_args()

    if args.verbose <= 1:
       loglevel = logging.ERROR

    ...

    if args.sub_commands == "download":
        download(args.target)
```
(also sub_command**s** should probably be sub_command?)

there could be a complication I don't understand though, so just a suggestion

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: Ah so there is implicit init here as well -- I think it would look neater to always call init() but make init() cope with the fact that initialization may already be done. The `download()` (or main if you call init from there) does not need to worry or print anything about initialization, it's all handled in `init()`

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: No exits in middle of app code IMO: you can just return

Also, you should expect errors in other tuf calls as well -- maybe you could just handle this once (even outside download() maybe?)

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: You don't know if it was fetched or already in cache. Maybe this print is not needed?

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: It may have been verified, failed to verify, or not even found... Maybe not a useful message?

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: this is not always correct  -- info.path is the url path which may or may not match the filename on disk. `path` contains the filesystem path

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: This should make it clear that we actually _downloaded_ the target

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: please review the printing in this function: I think it prints too much but more importantly the prints are incorrect (it always claims to have done a thing even if it did not do the thing at all). 

the docstring is incorrect: root.json is not downloaded, just copied from source dir (which is the correct choice in my opinion: we are giving a good example of shipping the root.json with the app)

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: This is a tricky one... how do we know it's 8000? what if that's in use already?

We can leave solving this to future PRs, but the magic string should not be hidden in this helper function: you have constants at the top of file already, maybe put this there as well for now?

If you define BASE_URL at the top of file, then this function is probably unneeded as it's a single call.

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: the directory name should be the app name so maybe 
`METADATA_DIR = f"{HOME_DIR}/.local/share/python-tuf-client-example"
`?

Also HOME_DIR seems to be defined just for this -- maybe the variable is not needed?

**Review by jku — CHANGES_REQUESTED** [review]
wow, ended up with a lot of comments -- if this is still WIP don't mind me

**Comment by jku** [comment]
This should probably be a future PR: we want the example tested by CI (should file a new issue for that)

For inspiration, I have some (quite bad) tests for a CLI app in https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tests/test_cli.py -- I think just a very simple test is fine: run a few commands, check if output contains correct keywords, check that files exist where they should

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: It was only a habit of using dictionaries for it. 🙂
As we discussed is more a choice, but I would like to use the namespaces as well.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I decided to return a boolean, but not sure if I should always return None for now.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I will review what I did, but insights are welcome.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: Please, confirm if it sounds ok.

**Comment by kairoaraujo** [comment]
I am still looking if ``download()`` could be more "elegant." More about the exception handling.

File: examples/client_example/README.md
Comment (jku) [inline examples/client_example/README.md]: I think "Python" is not the primary keyword we are interested in: either `Python-tuf Client Example` or `TUF Client Example` maybe? Same below where appropriate.

File: examples/client_example/README.md
Comment (jku) [inline examples/client_example/README.md]: The start is a bit weird for a section called "Repository"... Maybe something like 

```suggestion
The example client expects to find a TUF repository running on localhost. We can use the static metadata files in ``tests/repository_data/repository`` to set one up.
```

File: examples/client_example/README.md
Comment (jku) [inline examples/client_example/README.md]: Is this a useful paragraph?

File: examples/client_example/README.md
Comment (jku) [inline examples/client_example/README.md]: I wonder if we should just not include the output? I expect we'll tweak and improve the script, and then the README will be out of date, and anyone reading will be confused. 

Thoughts?

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: I think both of these prints are spammy (not needed): in 99.9% of the cases user is not interested, and will anyway get the message about trusted root later on.

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: type annotation

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: `info.path` is not a filesystem path. `path` is the filesystem path

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: `info.path` is not a filesystem path. `path` is the filesystem path

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: nitpick but does this now happen before e.g. "--help" is handled by argparse? Maybe init() should only be called from download (and the possible other top-level commands) so that --help or syntax errors would not create files and dirs?

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: nitpick: I guess the neatest way to handle this is to just call a `main()`  from here? this way any variables in the function are not made global

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: default is 0 so doesn't this mean a single "-v" does nothing?

**Review by jku — CHANGES_REQUESTED** [review]
I think the overall structure is good: client_args construction could be in its own function but so far it's so simple that I don't mind the current way.

Can you enable mypy for this? (currently still happens in setup.cfg I think)

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: This almost sounds like you could fit these five lines into a single line?

`"""Initialize local trusted metadata and create a directory for downloads"""`

or something like it

File: examples/client_example/client_example.py
Comment (jku) [inline examples/client_example/client_example.py]: Fixed a typo, made this more in line with the style guide:
```suggestion
    """Download the target file using ``ngclient`` Updater.
```

File: examples/client_example/README.md
Comment (kairoaraujo) [reply examples/client_example/README.md]: Visual in the source code landing page in Github is good. But I agree that it is not maintainable.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I thought it would be interesting to show the structure created on the client side.
As we show errors such as ``FileNotFound``, the user understands what is created in the local system.

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I also was dubious about it.
- Any new functionality (top-level commands) call init() inside
- Call init from main() as this is basic infrastructure for the client

I choose the second one, should I change?

Comment (jku) [reply examples/client_example/client_example.py]: it's a nitpick, so no need to change in this PR: just wanted to point that calling "--help" probably should not create files

File: examples/client_example/README.md
Comment (jku) [reply examples/client_example/README.md]: I think the PR makes my point as the output listed here does not seem in sync with the code :)

File: examples/client_example/client_example.py
Comment (kairoaraujo) [reply examples/client_example/client_example.py]: I wasn't clear in my comment.
I addressed the --help or syntax errors.

**Review by jku — APPROVED** [review]
I think this looks good: we'll definitely want to keep improving but this looks like a fine first step. The commit history could be squashed as it's a little random but since it's new code I don't really mind either way.

Anyone else want to have opinions here (@joshuagl, @sechkova)? 

Potential future work:
* Consider if we want the README published on RTD, modify it as required if we do
* Add a simple test for the example
* support other server locations than hard-coded localhost:8000
* support boostrapping root.json for other servers (like the repository generated by the repository example) somehow
* document the application decisions better (as an example _"why is is local metadata stored in $HOME/.local/share/"_, _"why not download first root.json over http?"_). I didn't do this in the review as I think it's a tricky balance: too much detail means no-one will read it...
* start thinking about how to use examples to show that TUF actually works, and prevents attacks. This might be more related to repository examples but ...

**Comment by jku** [comment]
I didn't actually run this before mysefl but now I did: I seem to get debug level output if i don't provide a verbosity level? This clearly seems like a bug?

**Comment by jku** [comment]
Merging this. 

Please file a new issue for adding a test for this example.

Maybe a new issue for the RTD documention inclusion as well? I'm not 100% sure we should include the example there but maybe someone should take a stab and see if it works well and whether it needs any changes.

**Pull request merged**
